### ISSUE TITLE
p2p shortener as a *client-side* feature/standard

### ISSUE THREAD

**Issue by thedod** [miguelfreitas/twister-html#262]
The need for p2p url-shortening is as old(?) as twister. We kept thinking of markups and "url payloads" to the twist's json, while it's as easy as inventing a language.

#### "Declaring" a link

* Simplest: just put it in a twist ([this twist](https://twisterio.com/post/thedod/2180) has 2, even)
* Declarative: `4:20 clock #url https://zzzen.com/420/` (only 1 per twist, also declares link text)

#### Refence:

* To a twist: `t://2180.thedod` would mean [this twist](https://twisterio.com/post/thedod/2180).
* To a url: `u://2180.thedod#1` would mean [the 2nd link](https://dubiousdod.org/go/UrlLengthening) in that twist. (`#0` can be omitted, of course. Usually 1 link per twist anyway)

#### Link text

* The default text of a `u://` link can be raw url with `...` for path: `https://dubiousdod.org...`
* If twist was of the form `^\s*(.*)\s+#url`\s+(<a valid url>)`\s*$` e.g. (`  ĎÚβĨŐÚŚ Dod   #url https://dubiousdod.org  `) then the first matched group is the text (and if it looks like a spoofed shortened URL e.g. `https://cnn.com...` - user should be warned with nasty css :tongue:)

**Comment by slr** [comment]
but who needs yet another language?

looks like declaring links as

    [text](link)

is a common practice.

and I suggest to not invent new URI scheme just for that reference. also why do you use inverted inheritage? I suppose

    @thedod/2180
    @thedod/2180#1

is enough if we want to have this.

**Comment by thedod** [comment]
:+1: I like this notation, because it covers both twists and links (of course, we should explicitly say `#0` to indicate it's the link and not the twist). If the `#n`th link is inside a `[text](url)` pattern, client knows what to use as link text. Otherwise, text can be an abbreviated url like `domain.tld/...`

**Comment by talexand** [comment]
Is consensus or traffic in support of this needed, or is it straight to PR?

**Comment by black-puppydog** [comment]
I would only support this if we raise the char limit quite a bit (2048 or above) because some links can be quite long and would still be shortened with 3rd party tools by the users.
That's exactly why we had that discussion in the first place. :/

**Comment by thedod** [comment]
"In real life"(TM), Tahoe-LAFS can create urls that are >140 chars (not by much, usually), and maybe there are systems that keep lots of important stuff in the query string, but I've yet to see a "real life" url longer than 242 chars, and `[A long url](`...242 chars...`)` is exactly 256 :wink:.

**Comment by black-puppydog** [comment]
Yeah, but if the urls take up any considerable amount (in percent) of the allowed chars then people have an incentive to use shorteners. Which is happening now and it would also happen when the char limit is 1000 and the links "only" 100.
I just think 3rd party shorteners are generally bad because [they break the internet](http://joshua.schachter.org/2009/04/on-url-shorteners).
Adding this functionality directly into twister would alleviate this since there would not be a point of failure/monitoring/censorship.
But if someone want to write a 900 char message with a 120 char url, they can and will shorten it, if twister provides that functionality or not.

**Comment by thedod** [comment]
@black-puppydog But that's *exactly* the purpose of this issue. Use a *twist* as an "unhosted shortening service" (especially now that @slr has [introduced markup](https://twisterio.com/post/tasty/2273)).
If the twist is long enough to hold [`some title`](`the url`) (i.e. 256, *maybe* 512), that twist *is* a shortener, and for *3* links:
* `@tasty/2273#0`:  [live](https://youtube.com/watch?v=bIoZMpu3w5Y)
* `@tasty/2273#1`: [cartoon](https://youtube.com/watch?v=xAwgsKlroFc)
* `@tasty/2273`: [@tasty/2273](https://twisterio.com/post/tasty/2273) [only we wouldn't need twisterio on a client :wink:)

Retraocactively, the entire twister history is a "shortener"
* `@thedod/2355` is [@thedod/2355](https://twisterio.com/post/thedod/2355)
* `@thedod/2355#0` is [www.youtube.com/...](https://www.youtube.com/watch?v=Tym0MObFpTI) (well, nobody would expect a sensible title for *that* link :laughing:)

**Comment by slr** [comment]
ok, URL shortener is coming finally — as a [daemon side feature](https://github.com/miguelfreitas/twister-core/commit/a00dfcf24724cae75e1a2df9bdeb7e35c61038d7). you may learn details and discuss it [there](https://groups.google.com/forum/#!topic/twister-dev/oDKUr9oOBHg).
